src/libraries/Microsoft.Extensions.Logging.Console/src/WindowsLogConsole.cs

```diff
@@ -17,42 +17,108 @@ public WindowsLogConsole(bool stdErr = false)
         }
 
         private bool SetColor(ConsoleColor? background, ConsoleColor? foreground)
+        {
+            var backgroundChanged = SetBackgroundColor(background);
+            return SetForegroundColor(foreground) || backgroundChanged;
+        }
+
+        public bool SetBackgroundColor(ConsoleColor? background)
         {
             if (background.HasValue)
             {
                 System.Console.BackgroundColor = background.Value;
+                return true;
             }
+            return false;
+        }
 
+        public bool SetForegroundColor(ConsoleColor? foreground)
+        {
             if (foreground.HasValue)
             {
                 System.Console.ForegroundColor = foreground.Value;
+                return true;
             }
-
-            return background.HasValue || foreground.HasValue;
+            return false;
         }
 
         private void ResetColor()
         {
             System.Console.ResetColor();
         }
 
-        public void Write(string message, ConsoleColor? background, ConsoleColor? foreground)
+        public void Write(ReadOnlySpan<char> span, ConsoleColor? background, ConsoleColor? foreground)
         {
-            bool colorChanged = SetColor(background, foreground);
-            _textWriter.Write(message);
+            var colorChanged = SetColor(background, foreground);
+            _textWriter.Write(span);
             if (colorChanged)
             {
                 ResetColor();
             }
         }
 
-        public void WriteLine(string message, ConsoleColor? background, ConsoleColor? foreground)
+        public void Write(string message)
         {
-            bool colorChanged = SetColor(background, foreground);
-            _textWriter.WriteLine(message);
-            if (colorChanged)
+            // writes content with the embedded foreground and background ansi colored. it skips non-color related ansi codes
+            (int startIndex, int length, ConsoleColor? bg, ConsoleColor? fg) content = (-1, 0, null, null);
```

Comment:
>content [](start = 77, length = 7)

is there any benefit to use Tuples here?

Reply:
for readability

Reply:
you have to access every tuple field later with `content.`. I'll leave it to you if you think it is more readable this way.

---
In reply to: [451228683](https://github.com/dotnet/runtime/pull/38616#discussion_r451228683) [](ancestors = 451228683)